File: services/calc/debrief_calc/tools/track/manipulation/generate_courses_speeds.py
"""Generate Courses and Speeds tool — derives course/speed from consecutive track positions."""

from __future__ import annotations

import math
from datetime import datetime
from typing import Any

from debrief_calc.models import ContextType, GeoJSONFeatureDict, SelectionContext, ToolCategoryEnum
from debrief_calc.registry import tool

EARTH_RADIUS_NM = 3440.065


def _haversine_distance(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """Calculate great-circle distance between two points in nautical miles."""
    lon1, lat1, lon2, lat2 = map(math.radians, [lon1, lat1, lon2, lat2])
    dlon = lon2 - lon1
    dlat = lat2 - lat1
    a = math.sin(dlat / 2) ** 2 + math.cos(lat1) * math.cos(lat2) * math.sin(dlon / 2) ** 2
    c = 2 * math.asin(math.sqrt(a))
    return c * EARTH_RADIUS_NM


def _initial_bearing(lon1: float, lat1: float, lon2: float, lat2: float) -> float:
    """Calculate initial bearing from point 1 to point 2 in degrees (0-360)."""
    lon1, lat1, lon2, lat2 = map(math.radians, [lon1, lat1, lon2, lat2])
    dlon = lon2 - lon1
    x = math.sin(dlon) * math.cos(lat2)
    y = math.cos(lat1) * math.sin(lat2) - math.sin(lat1) * math.cos(lat2) * math.cos(dlon)
    bearing = math.degrees(math.atan2(x, y))
    return (bearing + 360) % 360


def _parse_timestamp(iso_str: str) -> float:
    """Parse ISO8601 timestamp to seconds since epoch."""
    # Handle 'Z' suffix
    if iso_str.endswith("Z"):
        iso_str = iso_str[:-1] + "+00:00"
    dt = datetime.fromisoformat(iso_str)
    return dt.timestamp()


@tool(
    name="generate-courses-speeds",
    description=(
        "Derives course (bearing in degrees) and speed (knots) from consecutive "
        "track positions using Haversine distance and great-circle bearing formulas. "
        "Writes values into each position's course and speed fields, overriding any "
        "existing values."
    ),
    input_kinds=["TRACK"],
    output_kind="mutation/track/courses_speeds",
    context_type=ContextType.MULTI,
    category=ToolCategoryEnum.calc,
    parameters=[],
)
def generate_courses_speeds(
    context: SelectionContext, params: dict[str, Any]
) -> list[GeoJSONFeatureDict]:
    """Generate course and speed values for each position in track features.

    Args:
        context: SelectionContext with one or more track features.
        params: Empty dict (no parameters).

    Returns:
        List of modified track features with course/speed populated.
    """
    modified: list[GeoJSONFeatureDict] = []

    for feature in context.features:
        props = feature.get("properties", {})
        if props.get("kind") != "TRACK":
            continue

        coords = feature.get("geometry", {}).get("coordinates", [])
        positions = props.get("positions", [])
        n = len(positions)

        # Single-position track: return unchanged
        if n <= 1:
            modified.append(feature)
            continue

        # Phase 1: compute course and speed for each consecutive pair
        for i in range(n - 1):
            lon1, lat1 = coords[i][0], coords[i][1]
            lon2, lat2 = coords[i + 1][0], coords[i + 1][1]

            distance_nm = _haversine_distance(lon1, lat1, lon2, lat2)
            bearing = _initial_bearing(lon1, lat1, lon2, lat2)

            time1 = _parse_timestamp(positions[i]["time"])
            time2 = _parse_timestamp(positions[i + 1]["time"])
            elapsed_seconds = time2 - time1

            if distance_nm == 0:
                positions[i]["course"] = 0
                positions[i]["speed"] = 0
            elif elapsed_seconds <= 0:
                positions[i]["course"] = round(bearing, 2)
                positions[i]["speed"] = 0
            else:
                elapsed_hours = elapsed_seconds / 3600
                positions[i]["course"] = round(bearing, 2)
                positions[i]["speed"] = round(distance_nm / elapsed_hours, 2)

        # Phase 2: last position carries forward
        positions[n - 1]["course"] = positions[n - 2]["course"]
        positions[n - 1]["speed"] = positions[n - 2]["speed"]

        modified.append(feature)

    if not modified:
        raise ValueError("No track features found in input")

    return modified

File: apps/vscode/src/tools/track/manipulation/generateCoursesSpeeds.ts
/**
 * Generate Courses and Speeds tool implementation.
 *
 * Derives course (bearing) and speed (knots) from consecutive track positions
 * using Haversine distance and great-circle bearing formulas.
 */

import type { TrackFeature } from '@debrief/schemas';
import type { MCPToolDefinition } from '../../../types/tool';

const EARTH_RADIUS_NM = 3440.065;

function toRadians(deg: number): number {
  return (deg * Math.PI) / 180;
}

function toDegrees(rad: number): number {
  return (rad * 180) / Math.PI;
}

function haversineDistance(lon1: number, lat1: number, lon2: number, lat2: number): number {
  const lon1r = toRadians(lon1);
  const lat1r = toRadians(lat1);
  const lon2r = toRadians(lon2);
  const lat2r = toRadians(lat2);
  const dlon = lon2r - lon1r;
  const dlat = lat2r - lat1r;
  const a =
    Math.sin(dlat / 2) ** 2 + Math.cos(lat1r) * Math.cos(lat2r) * Math.sin(dlon / 2) ** 2;
  const c = 2 * Math.asin(Math.sqrt(a));
  return c * EARTH_RADIUS_NM;
}

function initialBearing(lon1: number, lat1: number, lon2: number, lat2: number): number {
  const lon1r = toRadians(lon1);
  const lat1r = toRadians(lat1);
  const lon2r = toRadians(lon2);
  const lat2r = toRadians(lat2);
  const dlon = lon2r - lon1r;
  const x = Math.sin(dlon) * Math.cos(lat2r);
  const y = Math.cos(lat1r) * Math.sin(lat2r) - Math.sin(lat1r) * Math.cos(lat2r) * Math.cos(dlon);
  const bearingDeg = toDegrees(Math.atan2(x, y));
  return ((bearingDeg % 360) + 360) % 360;
}

function round2(n: number): number {
  return Math.round(n * 100) / 100;
}

export const toolDefinition: MCPToolDefinition = {
  name: 'generate-courses-speeds',
  description:
    'Derives course (bearing in degrees) and speed (knots) from consecutive track positions ' +
    'using Haversine distance and great-circle bearing formulas. Writes values into each ' +
    "position's course and speed fields, overriding any existing values.",
  inputSchema: {
    type: 'object',
    properties: {
      features: {
        type: 'array',
        items: { type: 'object' },
        description: 'GeoJSON track features to process',
      },
    },
  },
  annotations: {
    'debrief:selectionRequirements': [{ kind: 'TRACK', min: 1 }],
    'debrief:category': 'track/manipulation',
    'debrief:version': '1.0.0',
    'debrief:outputKind': 'mutation/track/courses_speeds',
    'debrief:uiCategory': 'calc',
  },
};

interface Position {
  time: string;
  course?: number;
  speed?: number;
  [key: string]: unknown;
}

export function execute(features: TrackFeature[]): TrackFeature[] {
  const modified: TrackFeature[] = [];

  for (const feature of features) {
    const props = feature.properties ?? {};
    if (props.kind !== 'TRACK') {
      continue;
    }

    const coords = feature.geometry.coordinates as number[][];
    const positions = (props.positions as Position[]) ?? [];
    const n = positions.length;

    // Single-position track: return unchanged
    if (n <= 1) {
      modified.push(feature);
      continue;
    }

    // Phase 1: compute course and speed for each consecutive pair
    for (let i = 0; i < n - 1; i++) {
      const c1 = coords[i]!;
      const c2 = coords[i + 1]!;
      const lon1 = c1[0]!;
      const lat1 = c1[1]!;
      const lon2 = c2[0]!;
      const lat2 = c2[1]!;
      const pos = positions[i]!;
      const posNext = positions[i + 1]!;

      const distanceNm = haversineDistance(lon1, lat1, lon2, lat2);
      const bearing = initialBearing(lon1, lat1, lon2, lat2);

      const time1 = new Date(pos.time).getTime() / 1000;
      const time2 = new Date(posNext.time).getTime() / 1000;
      const elapsedSeconds = time2 - time1;

      if (distanceNm === 0) {
        pos.course = 0;
        pos.speed = 0;
      } else if (elapsedSeconds <= 0) {
        pos.course = round2(bearing);
        pos.speed = 0;
      } else {
        const elapsedHours = elapsedSeconds / 3600;
        pos.course = round2(bearing);
        pos.speed = round2(distanceNm / elapsedHours);
      }
    }

    // Phase 2: last position carries forward
    const last = positions[n - 1]!;
    const prev = positions[n - 2]!;
    last.course = prev.course;
    last.speed = prev.speed;

    modified.push(feature);
  }

  if (modified.length === 0) {
    throw new Error('No track features found in input');
  }

  return modified;
}
